sklearn/mixture/base.py

```diff
@@ -0,0 +1,454 @@
+
+        log_prob = self._estimate_log_prob(X)
+        return logsumexp(log_prob, axis=1)
+
+    def _initialize(self, X):
+        """Initialize the model parameters."""
+        resp = self._initialize_resp(X)
+        self._initialize_parameters(X, resp)
+
+    def _print_verbose_msg_init_beg(self, n_init):
+        """Print verbose message on initialization."""
+        if self.verbose == 1:
+            print("Initialization %d" % n_init)
+        elif self.verbose == 2:
+            print("Initialization %d" % n_init)
+            self._init_prev_time = time()
+            self._iter_prev_time = self._init_prev_time
+
+    def _print_verbose_msg_iter_end(self, n_iter, diff_ll):
+        """Print verbose message on initialization."""
+        if n_iter % self.verbose_interval == 0:
+            if self.verbose == 1:
+                print("  Iteration %d" % n_iter)
+            elif self.verbose == 2:
+                cur_time = time()
+                print("  Iteration %d\t time lapse %.5fs\t ll change %.5f" % (
+                    n_iter, cur_time - self._iter_prev_time, diff_ll))
+                self._iter_prev_time = cur_time
+
+    def _print_verbose_msg_init_end(self, ll):
+        """Print verbose message on the end of iteration."""
+        if self.verbose == 1:
+            print("Initialization converged: %s" % self.converged_)
+        elif self.verbose == 2:
+            print("Initialization converged: %s\t time lapse %.5fs\t ll %.5f" %
+                  (self.converged_, time() - self._init_prev_time, ll))
+
+    def _fit(self, X, y=None):
+        """Fit the model on data X.
+
+        The method fit the model `n_init` times and set the parameters with
+        which the model has the largest likelihood or lower bound. Within each
+        trial, the method iterates between E-step and M-step for `n_iter`
+        times until the change of likelihood or lower bound less than `tol`,
+        otherwise, a `ConvergenceWarning` is raised.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+
+        Returns
+        -------
+        self
+        """
+        X = check_X(X, self.n_components)
+        self.n_features = X.shape[1]
+        self._check_initial_parameters()
+
+        max_log_likelihood = -np.infty
+        best_params = self._get_parameters()
+
+        if self.verbose > 10:
+            self._log_snapshot = []
+
+        for init in range(self.n_init):
+            # print verbose msg
+            self._print_verbose_msg_init_beg(init)
+
+            self._initialize(X)
+            current_log_likelihood = -np.infty
+            self.converged_ = False
+
+            for i in range(self.n_iter):
+                prev_log_likelihood = current_log_likelihood
+
+                # e step
+                current_log_likelihood, resp = self._e_step(X)
+                change = abs(current_log_likelihood - prev_log_likelihood)
```

Comment:
I tested some cases with BayesianGaussianMixture. I think it might because of numerical issue when some mixtures starve (cannot cover enough data points) with 'tied' and precision, the change of LL would be negative. Other precision cases works fine. It might because 'tied' precision is not suitable for BayesianMixture